src/Orleans/Serialization/BuiltInTypes.cs

```diff
@@ -1637,7 +1637,7 @@ internal static object CopyDateTimeOffset(object obj)
 
         internal static void SerializeType(object obj, BinaryTokenStreamWriter stream, Type expected)
         {
-            stream.Write(((Type)obj).OrleansTypeName());
+            stream.Write(((Type)obj).OrleansTypeKeyString());
```

Comment:
It won't break backward compatibility in any case which works today (where `OrleansTypeKeyString()` & `OrleansTypeName()` just happen to produce the same result, eg:
```C#
> typeof(Grain).OrleansTypeName()
"Orleans.Grain"
> typeof(Grain).OrleansTypeKeyString()
"Orleans.Grain" // Simple (non-generic, non-array, etc) types not in "System" will be equal.
> typeof(int).OrleansTypeName()
"System.Int32"
> typeof(int).OrleansTypeKeyString()
"Int32" // Types in System wouldn't work.
```

If you look at `DeserializeType` you'll see it calls `ResolveTypeName` which looks for the type in `types`. If you look for where `types` gets modified, you'll see that the keys are generated using `OrleansTypeKeyString()` and not `OrleansTypeName()` - so the current serialize method is wrong to use `OrleansTypeName()`.